### PR TITLE
BUG: format small floats correctly (GH9764)

### LINKED ISSUE #9764: Displaying DataFrame with values from range [1e-7, 5e-7] shows 0 in some situations.
Hi,

For illustration purpose, I'll assume the following dataset:

```python

In [1]: d=pd.DataFrame({'col1':[9.999e-8, 1e-7, 1.0001e-7, 2e-7, 4.999e-7, 5e-7, 5.0001e-7, 6e-7]})

In [2]: d
Out[2]:
           col1
0  9.999000e-08
1  1.000000e-07
2  1.000100e-07
3  2.000000e-07
4  4.999000e-07
5  5.000000e-07
6  5.000100e-07
7  6.000000e-07
```

I've noticed the following behavior (in pandas 0.16 and 0.15.2):

1) When values from range [1e-7, 5e-7] are displayed along with values that are less than 1e-7, the output is OK:

```python

In [3]: d[0:6]
Out[3]:
           col1
0  9.999000e-08
1  1.000000e-07
2  1.000100e-07
3  2.000000e-07
4  4.999000e-07
5  5.000000e-07
```

2) When values exclusively from that range are displayed, the output is 0:

```python

In [4]: d[1:6]
Out[4]:
   col1
1     0
2     0
3     0
4     0
5     0
```

3) When values from that range are displayed along with values greater than 5e-7, the output for them is 0.000000:

```python

In [5]: d[1:8]
Out[5]:
       col1
1  0.000000
2  0.000000
3  0.000000
4  0.000000
5  0.000000
6  0.000001
7  0.000001
```

I'm assuming it's a rounding / output formatting error.

Also:

```python

In [6]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.4.2.final.0
python-bits: 64
OS: Linux
OS-release: 2.6.32-431.el6.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.16.0
nose: 1.3.4
Cython: 0.21.2
numpy: 1.9.2
scipy: 0.14.0
statsmodels: 0.6.1
IPython: 2.3.1
sphinx: None
patsy: 0.3.0
dateutil: 2.4.1
pytz: 2015.2
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.4.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
```

### PR DESCRIPTION
**Pull request by tomazberisa**
closes #9764 

In order to display a number from range (0,1) in decimal format with N significant digits (and not in scientific format), the number needs to be greater than or equal to 1e(-N+1), not 1e-N. 

Did not test for range of display.precision, as this does not seem to current practice.

Did not include tests as they would imply an output format which might change in the future.

### CODE REVIEW

**Comment by shoyer** [comment]
This should have a test to verify that your fix works. We use unit tests to verify our public API, but they're also for internal usage. The test will guarantee that nobody breaks this behavior without thinking carefully about it, even if we're not going to guarantee the output format to our users.

**Comment by tomazberisa** [comment]
Sure, test added to branch.

File: pandas/tests/test_format.py [hunk 1]
```diff
@@ -2986,6 +2986,29 @@ def test_format(self):
         self.assertEqual(result[0], " 12")
         self.assertEqual(result[1], "  0")
 
+    def test_output_significant_digits(self):
+        # Issue #9764
+        
+        # In case default display precision changes:
+        saved_option=pd.get_option('display.precision')
+        pd.set_option('display.precision', 7)
```
Comment (shoyer) [inline pandas/tests/test_format.py]: Can you set this with the context manager `pd.option_context()` instead? That's safer than setting a global variable.

File: pandas/tests/test_format.py [hunk 2]
```diff
@@ -2986,6 +2986,29 @@ def test_format(self):
...
+        
+        expected_output={
+            (0,6):'           col1\n0  9.999000e-08\n1  1.000000e-07\n2  1.000100e-07\n3  2.000000e-07\n4  4.999000e-07\n5  5.000000e-07',
+            (1,6):'           col1\n1  1.000000e-07\n2  1.000100e-07\n3  2.000000e-07\n4  4.999000e-07\n5  5.000000e-07',
+            (1,8):'           col1\n1  1.000000e-07\n2  1.000100e-07\n3  2.000000e-07\n4  4.999000e-07\n5  5.000000e-07\n6  5.000100e-07\n7  6.000000e-07',
+            (8,16):'            col1\n8   9.999000e-07\n9   1.000000e-06\n10  1.000100e-06\n11  2.000000e-06\n12  4.999000e-06\n13  5.000000e-06\n14  5.000100e-06\n15  6.000000e-06',
+            (9,16):'        col1\n9   0.000001\n10  0.000001\n11  0.000002\n12  0.000005\n13  0.000005\n14  0.000005\n15  0.000006'
+        }
+
+        for k, v in expected_output.items():
```
Comment (shoyer) [inline pandas/tests/test_format.py]: slightly cleaner to unpack the tuple: `for (start, stop), v in expected_output.items():`

File: pandas/tests/test_format.py [hunk 3]
```diff
@@ -2986,6 +2986,29 @@ def test_format(self):
...
+            self.assertEqual(d[k[0]:k[1]].__str__(), v)
```
Comment (shoyer) [inline pandas/tests/test_format.py]: Just use the `str()` function instead of calling `.__str__()`, e.g., `self.assertEqual(str(d[start:stop]), v)`

**Comment by shoyer** [comment]
looks pretty good -- could you add a line to the bug fix section of the release notes for v0.16.1?

**Comment by jreback** [comment]
also prob needs a test with mixed (meaning a really small number and a normal sized float). further this prob needs a quick way (eg vectorizes) to figure out whether to do this formatting (eg see how Timedelta does it for example) otherwise perf on a big array may be a problem

**Comment by tomazberisa** [comment]
@shoyer Thanks for comments - updated. I have also noticed that some of my commit msgs were not 100% in line with the contribution recommendations, so I've updated them.

**Comment by jreback** [comment]
ok this lgtm

**Comment by shoyer** [comment]
Thanks @tomazberisa!

**Pull request merged**